### PR TITLE
Fix a bunch of clippy lints (again)

### PR DESCRIPTION
**Pull request by coolreader18**
The biggest change in this is that I changed the representation of varargs in the bytecode and AST from `Option<Option<_>>` to its own enum in commit ab53883. I sorta jury-rigged it into the lalrpop file because I'm not too familiar with how it functions, so feel free to give critique for that.

### CODE REVIEW

File: vm/src/bytecode.rs [hunk 1]
```diff
@@ -401,3 +408,23 @@ impl fmt::Debug for CodeObject {
...
+        match varargs {
+            ast::Varargs::None => Varargs::None,
+            ast::Varargs::NoCapture => Varargs::NoCapture,
+            ast::Varargs::Capture(param) => Varargs::Capture(param.arg),
+        }
+    }
+}
+
+impl Varargs {
+    pub fn from_ast_vararg_ref(varargs: &ast::Varargs) -> Varargs {
```
Comment (OddCoincidence) [inline vm/src/bytecode.rs]: Why not `impl<'a> From<&'a ast::Varargs> for Varargs`?

File: vm/src/obj/objlist.rs [hunk 2]
```diff
@@ -203,7 +203,7 @@ impl PyListRef {
                 return Ok(index);
             }
         }
-        let ref needle_str = vm.to_str(&needle)?.value;
+        let needle_str = vm.to_pystr(&needle)?;
```
Comment (OddCoincidence) [inline vm/src/obj/objlist.rs]: Was removing the `.value` intentional here? Will the `Display` impl for `PyString{,Ref}` do the right thing here?

File: vm/src/obj/objtype.rs [hunk 3]
```diff
@@ -331,6 +331,7 @@ fn linearise_mro(mut bases: Vec<Vec<PyClassRef>>) -> Option<Vec<PyClassRef>> {
     Some(result)
 }
 
+#[allow(clippy::implicit_hasher)]
```
Comment (OddCoincidence) [inline vm/src/obj/objtype.rs]: We could probably disable this one at the crate level, I don't think it's useful at all

File: vm/src/bytecode.rs (hunk 1 above)
Comment (coolreader18) [reply vm/src/bytecode.rs]: I tried that, but `.into()` didn't work in `compile.rs`, so I sorta just gave up. It does work manually writing `from(&args.varargs)`, though, so I'll change it to be this.

File: vm/src/pyobject.rs [hunk 4]
```diff
@@ -1118,9 +1118,14 @@ impl PyObject {
     }
 
     #[inline]
-    pub fn payload<T: PyValue>(&self) -> Option<&T> {
+    pub fn payload<T: PyObjectPayload>(&self) -> Option<&T> {
         self.payload.as_any().downcast_ref()
     }
+
+    #[inline]
+    pub fn payload_is<T: PyObjectPayload>(&self) -> bool {
```
Comment (OddCoincidence) [inline vm/src/pyobject.rs]: nit: would prefer this just be called `is`

(I'm thinking once there's a `downcast` or `as` on `PyRef`, we'll move away from using `payload`, so calling this `is` would provide a consistent api)

File: vm/src/obj/objlist.rs (hunk 2 above)
Comment (coolreader18) [reply vm/src/obj/objlist.rs]: Oh, looking back on it it's probably not the best thing to do; thanks for pointing it out.

File: parser/src/ast.rs [hunk 5]
```diff
@@ -391,3 +391,28 @@ pub enum StringGroup {
...
+    Capture(Parameter),
+}
+
+impl Default for Varargs {
+    fn default() -> Varargs {
+        Varargs::None
+    }
+}
+
+impl From<Option<Option<Parameter>>> for Varargs {
```
Comment (OddCoincidence) [inline parser/src/ast.rs]: Where's the double `Option` coming from? Is there a way to eliminate it altogether?

Comment (coolreader18) [reply parser/src/ast.rs]: I didn't really want to mess around in the `python.lalrpop` grammar, as I don't really have any experience with it, so I opted for this instead.

Comment (OddCoincidence) [reply parser/src/ast.rs]: Okay, that's fine, was just wondering

File: vm/src/pyobject.rs (hunk 4 above)
Comment (OddCoincidence) [reply vm/src/pyobject.rs]: Oh, bummer, either is fine for now then

File: parser/src/ast.rs (hunk 5 above)
Comment (windelbouwman) [reply parser/src/ast.rs]: The double option is a sign of a triple option enum. Actually the options are: namedparameter, namedstar, and unnamed star.

File: parser/src/ast.rs [hunk 6]
```diff
@@ -391,3 +391,28 @@ pub enum StringGroup {
         values: Vec<StringGroup>,
     },
 }
+
+#[derive(Debug, PartialEq)]
+pub enum Varargs {
+    None,
+    NoCapture,
```
Comment (windelbouwman) [inline parser/src/ast.rs]: Suggestion to rename this into UnNamed and Named.

Comment (coolreader18) [reply parser/src/ast.rs]: Sounds good.

**Pull request merged**